File: klox/src/main/kotlin/dev/craigfurman/klox/Expression.kt
package dev.craigfurman.klox

sealed interface Expression {
    data class Assign(val name: Token, val value: Expression) : Expression

    data class Binary(
        val left: Expression,
        val operator: Token,
        val right: Expression
    ) : Expression

    data class Call(
        val callee: Expression,
        val paren: Token,
        val arguments: List<Expression>
    ) : Expression

    data class Get(val obj: Expression, val name: Token) : Expression
    data class Grouping(val expr: Expression) : Expression
    data class Literal(val value: Any?) : Expression

    data class Logical(
        val left: Expression,
        val operator: Token,
        val right: Expression
    ) : Expression

    data class SetExpr(
        val obj: Expression,
        val name: Token,
        val value: Expression
    ) : Expression

    data class Super(val keyword: Token, val method: Token) : Expression
    data class This(val keyword: Token) : Expression
    data class Unary(val operator: Token, val right: Expression) : Expression
    data class Variable(val name: Token) : Expression

    // The book uses a more-traditional visitor pattern, with an accept(Visitor)
    // method defined on the Expression interface. It seemed a shame to generate
    // all the boilerplate for that, and I wanted to try when...is statements
    // with this sealed class. I might measure the performance and see how it
    // compares to the proper visitor pattern later.
    interface Visitor<R> {
        fun visitAssignExpr(expr: Assign): R
        fun visitBinaryExpr(expr: Binary): R
        fun visitCallExpr(expr: Call): R
        fun visitGetExpr(expr: Get): R
        fun visitGroupingExpr(expr: Grouping): R
        fun visitLiteralExpr(expr: Literal): R
        fun visitLogicalExpr(expr: Logical): R
        fun visitSetExpr(expr: SetExpr): R
        fun visitSuperExpr(expr: Super): R
        fun visitThisExpr(expr: This): R
        fun visitUnaryExpr(expr: Unary): R
        fun visitVariableExpr(expr: Variable): R

        fun visit(expr: Expression) = when (expr) {
            is Assign -> visitAssignExpr(expr)
            is Binary -> visitBinaryExpr(expr)
            is Call -> visitCallExpr(expr)
            is Get -> visitGetExpr(expr)
            is Grouping -> visitGroupingExpr(expr)
            is Literal -> visitLiteralExpr(expr)
            is Logical -> visitLogicalExpr(expr)
            is SetExpr -> visitSetExpr(expr)
            is Super -> visitSuperExpr(expr)
            is This -> visitThisExpr(expr)
            is Unary -> visitUnaryExpr(expr)
            is Variable -> visitVariableExpr(expr)
        }
    }
}

File: golox/lox/expression.go
package lox

type Expr interface {
	Accept(visitor ExprVisitor) (any, error)
}

type ExprVisitor interface {
	VisitAssignExpr(expr *AssignExpr) (any, error)
	VisitBinaryExpr(expr *BinaryExpr) (any, error)
	VisitCallExpr(expr *CallExpr) (any, error)
	VisitGetExpr(expr *GetExpr) (any, error)
	VisitGroupingExpr(expr *GroupingExpr) (any, error)
	VisitLiteralExpr(expr *LiteralExpr) (any, error)
	VisitLogicalExpr(expr *LogicalExpr) (any, error)
	VisitSetExpr(expr *SetExpr) (any, error)
	VisitThisExpr(expr *ThisExpr) (any, error)
	VisitUnaryExpr(expr *UnaryExpr) (any, error)
	VisitVarExpr(expr *VarExpr) (any, error)
}

type AssignExpr struct {
	name Token
	expr Expr
}

type BinaryExpr struct {
	left     Expr
	operator Token
	right    Expr
}

type CallExpr struct {
	callee    Expr
	paren     Token
	arguments []Expr
}

type GetExpr struct {
	obj  Expr
	name Token
}

type GroupingExpr struct {
	expr Expr
}

type LiteralExpr struct {
	value any
}

type LogicalExpr struct {
	left     Expr
	operator Token
	right    Expr
}

type SetExpr struct {
	obj   Expr
	name  Token
	value Expr
}

type ThisExpr struct {
	keyword Token
}

type UnaryExpr struct {
	operator Token
	right    Expr
}

type VarExpr struct {
	name Token
}

func (e *AssignExpr) Accept(visitor ExprVisitor) (any, error)   { return visitor.VisitAssignExpr(e) }
func (e *BinaryExpr) Accept(visitor ExprVisitor) (any, error)   { return visitor.VisitBinaryExpr(e) }
func (e *CallExpr) Accept(visitor ExprVisitor) (any, error)     { return visitor.VisitCallExpr(e) }
func (e *GetExpr) Accept(visitor ExprVisitor) (any, error)      { return visitor.VisitGetExpr(e) }
func (e *GroupingExpr) Accept(visitor ExprVisitor) (any, error) { return visitor.VisitGroupingExpr(e) }
func (e *LiteralExpr) Accept(visitor ExprVisitor) (any, error)  { return visitor.VisitLiteralExpr(e) }
func (e *SetExpr) Accept(visitor ExprVisitor) (any, error)      { return visitor.VisitSetExpr(e) }
func (e *ThisExpr) Accept(visitor ExprVisitor) (any, error)     { return visitor.VisitThisExpr(e) }
func (e *LogicalExpr) Accept(visitor ExprVisitor) (any, error)  { return visitor.VisitLogicalExpr(e) }
func (e *UnaryExpr) Accept(visitor ExprVisitor) (any, error)    { return visitor.VisitUnaryExpr(e) }
func (e *VarExpr) Accept(visitor ExprVisitor) (any, error)      { return visitor.VisitVarExpr(e) }
